Add test to validate server attributes of "qstat -Bf"

npadole20 opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Bug/feature Description

* Add test case to validate server attributes in the output of qstat -Bf, as root and non-root user

#### Affected Platform(s)
* All

#### Cause / Analysis / Design
* This test verifies that the values of server attributes reported by qstat -Bf matches the expected default values. 

#### Solution Description
* Value of server attributes reported in qstat -Bf is checked against the expected default value of that attribute. Some attribute values are changes in the setup function of PTL, those attributes are checked against the new values set by PTL.

#### Testing logs/output
* Attached in comments

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [ ] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [ ] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [ ] I have attached **test logs to this pull request** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

npadole20 commented:
[test_qstat_Bf_as_root_final.txt](https://github.com/PBSPro/pbspro/files/2265981/test_qstat_Bf_as_root_final.txt)
[test_qstat_Bf_as_user_final.txt](https://github.com/PBSPro/pbspro/files/2265982/test_qstat_Bf_as_user_final.txt)
[test_qstat_Bf_as_root_final_cray.txt](https://github.com/PBSPro/pbspro/files/2265983/test_qstat_Bf_as_root_final_cray.txt)
[test_qstat_Bf_as_user_final_cray.txt](https://github.com/PBSPro/pbspro/files/2265984/test_qstat_Bf_as_user_final_cray.txt)
[test_qstat_Bf_as_root_sim.txt](https://github.com/PBSPro/pbspro/files/2265989/test_qstat_Bf_as_root_sim.txt)
[test_qstat_Bf_as_user_sim.txt](https://github.com/PBSPro/pbspro/files/2265990/test_qstat_Bf_as_user_sim.txt)

anamikau commented on test/tests/functional/pbs_qstat_validate_attr.py:
```diff
@@ -0,0 +1,117 @@
+        attr_dict = {}
+        command = os.path.join(self.server.pbs_conf['PBS_EXEC'],
+                               'bin', 'qstat -Bf')
+        ret = self.du.run_cmd(self.server.hostname, command)
```
any reason why are you not using server.status

npadole20 replied:
thank you for this suggestion. Helped in reducing lines of code.

anamikau commented on test/tests/functional/pbs_qstat_validate_attr.py:
```diff
@@ -0,0 +1,117 @@
+        attr_dict = {}
+        command = os.path.join(self.server.pbs_conf['PBS_EXEC'],
+                               'bin', 'qstat -Bf')
+        ret = self.du.run_cmd(self.server.hostname, command, sudo=True)
```
same, can we use server.status

anamikau commented on test/tests/functional/pbs_qstat_validate_attr.py:
```diff
@@ -0,0 +1,117 @@
+                        out1.append(k.split(':')[0])
+                    self.assertListEqual(qstat_o[each], out1)
+                else:
+                    self.assertEqual(qstat_o[each], attr_dict[each])
```
please add log messages. right now test output just prints following

2018-08-07 01:13:37,722 INFO     =====================================
2018-08-07 01:13:37,723 INFO     Completed TestQstatAttrValidate setUp
2018-08-07 01:13:37,723 INFO     =====================================
2018-08-07 01:13:37,723 INFOCLI2 x31-64-sles11sp2: sudo -H /opt/pbs/bin/qstat -Bf

bhroam commented:
There is a similar test in SmokeTest for nodes.  Should this one be there as well?

anamikau commented:
Hi @bhroam , I was going to suggest to add smoke tag for this test. we can move this to smokeTest but that file is already too big. I suggest add smoke tag and keeping either this in a separate file or may be move to https://github.com/PBSPro/pbspro/blob/master/test/tests/functional/pbs_qstat_tag.py, but update the class name and file name to be generic to something like pbs_qstat.

npadole20 commented:
@anamikau  [https://github.com/PBSPro/pbspro/blob/master/test/tests/functional/pbs_qstat_json.py](https://github.com/PBSPro/pbspro/blob/master/test/tests/functional/pbs_qstat_json.py) also has test cases which validate qstat output, although it is done for json format.

anamikau commented on test/tests/functional/pbs_qstat_validate_attr.py:
```diff
@@ -73,45 +81,34 @@ def validate_el(self, attr_dict, user):
+                    self.assertEqual(qstat_o[each], attr_dict[each],
+                                     "Value of server attribute \"%s\""
+                                     "does not match expected value" % each)
+        self.logger.info(
```
Please also add a log message inside the loop printing what values are successfully matched. for example, self.logger.info ("successfully verified that qstat_o[each] is same as attr_dict[each]")

anamikau commented:
@npadole20 , I like your suggestion for moving this under https://github.com/PBSPro/pbspro/blob/master/test/tests/functional/pbs_qstat_json.py. Please rename pbs_qstat_json.py to pbs_qstat.py and also the class name accordingly.

anamikau commented:
Also please add smoke tag to this test.

npadole20 commented:
Accidentally closed the PR. Reopened it too. 
@anamikau thank you for the review. I have addressed all your review comments. 
@bhroam I hope the latest changes address your comment.

anamikau commented:
Also can you provide the latest test execution results.

bhroam commented on test/tests/functional/pbs_qstat.py:
```diff
@@ -57,6 +58,61 @@ def parse_json(self, dictitems, qstat_attr):
+                else:
+                    self.assertEqual(qstat_o[each], attr_dict[each],
+                                     "Value of server attribute \"%s\""
+                                     "does not match expected value" % each)
```
I think this doesn't quite do the right thing.  This will make sure the correct attributes are in the qstat output, but if an unexpected attribute is there, it will mark that as successfully validated.  I think what you want is to replace the if each in qstat_o with an self.assertIn(each, qstat_o).

npadole20 replied:
There are some attributes from qstat -Bf which we are not checking in the test case like flicenses, pbs_license_info. In that case, self.assertIn(each, qstat_o) will throw an exception. To avoid this, we can add all possible attributes in output of "qstat -Bf" to 'qstat_o' and set the value as 'None' for attributes which we don't want to validate.

bhroam commented on test/tests/functional/pbs_qstat.py:
```diff
@@ -57,6 +58,61 @@ def parse_json(self, dictitems, qstat_attr):
+        qstat_o['eligible_time_enable'] = 'False'
+        qstat_o['max_concurrent_provision'] = '5'
+
+        if user == ROOT_USER:
```
This isn't quite true.  What you want to know is if the user is a manager.  The fact that more attributes show up is that root is a manager.  This is probably good enough though.

bhroam commented on test/tests/functional/pbs_qstat.py:
```diff
@@ -57,6 +58,61 @@ def parse_json(self, dictitems, qstat_attr):
+            qstat_o['restrict_res_to_release_on_suspend'] = 'ncpus'
+
+        for each in attr_dict:
+            if each in qstat_o:
```
This won't determine if there is an attribute missing from attr_dict that is in qstat_o.

bhroam commented on test/tests/functional/pbs_qstat.py:
```diff
@@ -94,22 +99,27 @@ def validate_el(self, attr_dict, user):
+                                 " is validated successfully" % each)
+        value = { k : qstat_o[k] for k in set(qstat_o) - set(attr_dict) }
+        self.assertEqual(len(value), 0, "Output of qstat -Bf is missing"
+                                        " some expected attributes - %s" % value.keys())
```
Since you only want the keys from the value dictionary, you could just create a list instead of a dictionary with your list comprehension.

bhroam commented on test/tests/functional/pbs_qstat.py:
```diff
@@ -94,22 +99,27 @@ def validate_el(self, attr_dict, user):
-            if each in qstat_o:
+            self.assertIn(each, qstat_o, "Unexpected attribute \"%s\""
+                          " encountered in output of qstat -Bf" %each)
+            if qstat_o[each] != None:
```
Say 'is not None' instead of '!= None'

bhroam commented on test/tests/functional/pbs_qstat.py:
```diff
@@ -94,22 +99,27 @@ def validate_el(self, attr_dict, user):
         for each in attr_dict:
-            if each in qstat_o:
+            self.assertIn(each, qstat_o, "Unexpected attribute \"%s\""
+                          " encountered in output of qstat -Bf" %each)
```
Could you add a space after the %?

bhroam commented:
Travis failed to pep8.

npadole20 commented:
[test_qstat_Bf_as_root_B_2.txt](https://github.com/PBSPro/pbspro/files/2296965/test_qstat_Bf_as_root_B_2.txt)
[test_qstat_Bf_as_user_B_2.txt](https://github.com/PBSPro/pbspro/files/2296966/test_qstat_Bf_as_user_B_2.txt)

npadole20 commented:
The travis error is -

2018-08-17 11:23:05,257 INFO     ======================================================================
2018-08-17 11:23:05,257 INFO     FAILED: test_qstat_Bf_as_user (tests.functional.pbs_qstat.TestQstat)

2018-08-17 11:23:05,257 INFO     ___m_oo_m___
2018-08-17 11:23:05,257 INFO     Traceback (most recent call last):
  File "/home/travis/build/PBSPro/pbspro/test/tests/functional/pbs_qstat.py", line 310, in test_qstat_Bf_as_user
    self.validate_el(attr_dict, TEST_USER)
  File "/home/travis/build/PBSPro/pbspro/test/tests/functional/pbs_qstat.py", line 127, in validate_el
    % value)
AssertionError: Output of qstat -Bf is missing some expected attributes - ['pbs_license_info']

But I don't get this error while executing the test case on the test machine.

bhroam commented on test/tests/functional/pbs_qstat.py:
```diff
@@ -117,9 +117,10 @@ def validate_el(self, attr_dict, user):
                 self.logger.info("Server attribute \"%s\""
                                  " is validated successfully" % each)
-        value = { k : qstat_o[k] for k in set(qstat_o) - set(attr_dict) }
+        value = [k for k in qstat_o.keys() if k not in attr_dict.keys()]
```
Just as a note, you don't need to use .keys().  When you say x in dict, you're matching a key.  This will be slightly slower since you have to create a list when you don't really need to, but not overly so.  Change it if you want.

bhroam commented:
I'm not 100% sure why travis failed on most of the runs.  I restarted the builds.

bhroam commented:
Your issue is that pbs_license_info is not set by default.  It is set by our internal tools before the test is run.

npadole20 commented:
@bhroam Thanks for the review. 
Also, can this PR be merged even if Travis is failing?

npadole20 commented on test/tests/functional/pbs_qstat.py:
```diff
@@ -57,6 +58,76 @@ def parse_json(self, dictitems, qstat_attr):
+        qstat_o['license_count'] = None
+        qstat_o['managers'] = None
+        qstat_o['flatuid'] = None
+        qstat_o['id'] = self.server.shortname
```
It's not reported by default in qstat -Bf. I caught that when I integrated Bhroam's review comments. Earlier, self.assertEqual(qstat_o[each], attr_dict[each]) was not catching the fact that 'scheduler_iteration' is missing from 'attr_dict'

anamikau replied:
Ah ok I see. Thanks.

anamikau commented:
Code change still looks good to me.

hirenvadalia commented on test/tests/functional/pbs_qstat.py:
```diff
@@ -57,6 +58,75 @@ def parse_json(self, dictitems, qstat_attr):
+        qstat_o['id'] = self.server.shortname
+
+        # If test case is run as root, PTL sets acl_roots
+        if os.getuid() == 0:
```
Just a suggestion take it leave it up to you: You can merge this if condition with below if condition

npadole20 replied:
But there is a difference. os.getuid() is in the case when the test case is being run as root. The other if condition is to check if qstat -Bf is executed as root.

hirenvadalia replied:
Ah you are right! you need separate if condition

hirenvadalia commented on test/tests/functional/pbs_qstat.py:
```diff
@@ -57,6 +58,75 @@ def parse_json(self, dictitems, qstat_attr):
                         self.parse_json(val, qstat_attr)
         return qstat_attr
 
+    def validate_el(self, attr_dict, user):
```
Just a suggestion take it leave it up to you: For more readability I suggest to change this method name to validate_attribute, what is `el`? 😄

hirenvadalia commented on test/tests/functional/pbs_qstat.py:
```diff
@@ -228,3 +298,21 @@ def test_qstat_qf_json_valid_multiple_queues(self):
+        Validate output of qstat -Bf when executed as non-root user
+        """
+        ret = self.server.status(runas=TEST_USER)
+        attr_dict = ret[0]
```
Do you really need this variable? you can directly pass `ret[0]` in below call

npadole20 commented:
After discussion with @anamikau,I am closing this PR and will raise it in extensions.
